### PR TITLE
Resolve issue with implicit yields requiring Zero

### PR DESCRIPTION
**Pull request by laenas**
Related to #8587 

I've been wrapping my head around this in the idle hours for a couple of days now and this seems like it should resolve the bug in question.  But I would appreciate eyes on it, and will with great joy take guidance on more elegant means of solving this.

Here's the low-down on the bug:
Based on [the RFC](https://github.com/fsharp/fslang-design/blob/13509be8a445d0b75d63ed9f44f9f127c21bcc51/FSharp-4.7/FS-1069-implicit-yields.md) - it's expected that a yielding CE require Zero (see Detailed Design - point 2)

In the code itself, we can see that the behavior makes sense:
CheckComputationExpressions.fs : 1440 states
> // "expr;" in final position is treated as { expr; zero }

And thus, relative to the samples in the ticket - the final 3, a raw expression, would be followed by a zero which would somewhat necessitate the need for that Zero method to exist.

The resolution for this that I have found is to do a safety-check with regards to how we handle ImplicitZero - if we're in a context where Implicit Yielding is enabled, then rather than erroring out if the builder lacks Zero, we can first check to see whether or not we're in an Implicit Yielding situation and therefore, if we are, rely on the translation context provided to handle the zero intelligently.

I've considered the placement of this logic - but cannot identify a better place to shim it in (without affecting more of the surrounding code than I feel comfortable with) than in `tryTrans` - while this results in a duplicate check around the `enableImplicitYield` value (now in `tryTrans` and then nearly immediately afterwards in in `translatedCtxt` callback defined at :1452) - this seems minor.  

My intuition is that there's a cleaner way to handle this, but it may require more significant modifications to the surrounding behaviors here than what I have made here and, as such, I hesitate to do so without further insights.

### CODE REVIEW

**Comment by cartermp** [comment]
Since this would require conformance testing, here is a good location:

https://github.com/dotnet/fsharp/tree/20dec2af41d075a71a04f92c15a09a9a38c4752f/tests/FSharp.Compiler.ComponentTests/Language

Sadly tests are hard to make sense out of as various conformance stuff is strewn about in different projects and suites.

I would add these for now:

* Test with implicit yield with no zero - none defined in the builder - expected pass
* Test with implicit yield with a zero - none defined in the builder - expected fail
* Test with implicit yield with a zero - zero defined in the builder - expected pass

Note that you can write the tests like so:

```fsharp
[<Fact>]
let ``test-name``() =
     FSharp """
// the F# source code
    """
    |> compile
    |> shouldSucceed
```

or  (perhaps preferred, up to you)

```fsharp
[<Fact>]
let ``test-name``() =
     FSharp """
// the F# source code
    """
    |> typecheck
    |> shouldSucceed
```

or 


```fsharp
[<Fact>]
let ``test-name``() =
     FSharp """
// the F# source code
    """
    |> typecheck
    |> shouldFail
    |> withDiagnostics [ (Error 0, Line 0, Col 0, Line 0, Col 0, "the-message" ]
```

Where the error is an error code, and the line/col numbers correspond to start/end ranges for the error, and the string message is the compiler message.

alternatively, the `withDiganostics` could be elided, but I think we'd prefer if they were included

**Comment by cartermp** [comment]
From here it'll be easier to iterate on edge cases if you think of them

**Comment by laenas** [comment]
Perfect - I'll get some tests wrapped around this and update the PR!

**Comment by laenas** [comment]
Beyond the three obvious cases, I also added in the check for the implicit zeroing that occurs with if/else - if that's overkill or redundant (as I could see it being more a matter for how the parser generates the synexpr tree), just let me know.

**Review by cartermp — APPROVED** [review]
Thanks!

File: src/fsharp/CheckComputationExpressions.fs [hunk 1]
```diff
@@ -950,7 +950,8 @@ let TcComputationExpression cenv env overallTy tpenv (mWhole, interpExpr: Expr,
             error(Error(FSComp.SR.tcConstructIsAmbiguousInComputationExpression(), m))
 
         | SynExpr.ImplicitZero m -> 
-            if isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env m ad "Zero" builderTy) then error(Error(FSComp.SR.tcRequireBuilderMethod("Zero"), m))
+            if (not enableImplicitYield) && 
+               isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env m ad "Zero" builderTy) then error(Error(FSComp.SR.tcRequireBuilderMethod("Zero"), m))
             Some (translatedCtxt (mkSynCall "Zero" m []))
```
Comment (dsyme) [inline src/fsharp/CheckComputationExpressions.fs:955]: Could you explain more - what happens to the invalid ` (mkSynCall "Zero" m [])` in the case where `enableImplicitYield` is true?  (invalid because the method may not exist).

Why doesn't `translatedCtxt` end up using this expression ?

Comment (dsyme) [reply src/fsharp/CheckComputationExpressions.fs:955]: I see, in the case of implicit yields the holeFill is discarded here https://github.com/dotnet/fsharp/blob/ac4d5471c46fe1036a6a55214275f3f3f75daf93/src/fsharp/CheckComputationExpressions.fs#L1456

I need to think about this briefly

Comment (dsyme) [reply src/fsharp/CheckComputationExpressions.fs:955]: So that holeFill is only discarded when the StatementOrYield thing fails to check as a statement.

So please add test cases for the cases where the last thing is a statement, e.g.

```
let x = 
    [ 1; 
      if true then 2
      printfn "hello" ]
```
In these cases we do need a Zero, and a good error message if it's missing.  I suspect it will generate an error message but a little later as it checks the expression returned by `mkSynCall` , which is actually ok tbh

Comment (dsyme) [reply src/fsharp/CheckComputationExpressions.fs:955]: OK I see you have this test case here: https://github.com/dotnet/fsharp/pull/10556/files/ac4d5471c46fe1036a6a55214275f3f3f75daf93#diff-ac5756e06cf81b30d7554703b9cd9cd75c87e327fb985dd262e01696e57a0426R41

and indeed it returns a different error message.

Then this is all ok.  It would be good to add a comment along the lines 

```
// In some cases the node produced by `mkSynCall "Zero" m []` may be discarded in the case
// of implicit yields - for example "list { 1; 2 }" when each expression checks as an implicit yield.
// If it is not discarded, the syntax node will later be checked and the existence/non-existence of the Zero method
// will be checked/reported appropriately (though the error message won't mention computation expressions
// like our other error messages for missing methods).
```

**Review by dsyme — APPROVED** [review]
Yes, this is a good change, thank you

File: src/fsharp/CheckComputationExpressions.fs (hunk 1 above)
Comment (laenas) [reply src/fsharp/CheckComputationExpressions.fs:955]: It sounds like you've answered your first question:  This seems like a weird gap in implementation - because yeah, we're using that passed-in `translatedCtxt` to deal with the implicitly yielded Zero being transformed out - but that check for the Zero method stops us before we get there.

With regards to the testing, I suspect the error message in-place should work comfortably enough:
> "The type 'ListBuilder' does not define the field, constructor or member 'Zero'."

With that said, is another test strictly needed?  Part of the testing above handles the case of a direct unit return as the final implicitly yielded element - both in the positive and negative cases.  Would the call to a unit-producing function like printfn change the resolution semantics?

Comment (laenas) [reply src/fsharp/CheckComputationExpressions.fs:955]: Oh, the joys of Github not updating comments while I'm writing and doing some double-checking on my work :)
I'll add said comment presently!
